File: core/src/main/scala/com/sksamuel/scrimage/RGBColor.scala
package com.sksamuel.scrimage

/** @author Stephen Samuel */
trait Color {
  /**
   * Returns a conversion of this Color into an RGBColor.
   * If this colour is already an instance of RGBColor then the same instance will be returned.
   *
   * @return an RGBColor conversion of this color.
   */
  def toRGB: RGBColor
}

object Color {
  implicit def color2rgb(color: Color): RGBColor = color.toRGB
  implicit def color2awt(color: Color): java.awt.Color = new java.awt.Color(color.toRGB.rgba)
  implicit def awt2color(awt: java.awt.Color): RGBColor = RGBColor(awt.getRed, awt.getBlue, awt.getGreen, awt.getAlpha)
}

case class RGBColor(red: Int, green: Int, blue: Int, alpha: Int = 0) extends Color {
  require(0 <= red && red <= 255, "Red component is invalid")
  require(0 <= green && green <= 255, "Green component is invalid")
  require(0 <= blue && blue <= 255, "Blue component is invalid")
  require(0 <= alpha && alpha <= 255, "Alpha component is invalid")

  def toRGB: RGBColor = this

  /**
   * Returns as an int the value of this color. The RGB and alpha components are packed
   * into the int as byes.
   * @return
   */
  def rgba: Int = red & 0xFF | green >> 8 & 0xFF | blue >> 16 & 0xFF | alpha >> 24 & 0xFF

  /**
   * Returns a HEX String of this colour. Eg for 0,255,0, this method will return 00FF00.
   */
  def toHex: String = String.format("%02x%02x%02x", red, green, blue)
}

/**
 * Also called HSB.
 *
 * The hue component should be between 0.0 and 360.0
 * The saturation component should be between 0.0 and 1.0
 * The lightness component should be between 0.0 and 1.0
 * The alpha component should be between 0.0 and 1.0
 *
 * @param hue
 * @param saturation
 * @param value
 */
case class HSVColor(hue: Float, saturation: Float, value: Float, alpha: Float) extends Color {
  require(0 <= hue && hue <= 360f, "Hue component is invalid")
  require(0 <= saturation && saturation <= 1f, "Saturation component is invalid")
  require(0 <= value && value <= 1f, "Value component is invalid")
  require(0 <= alpha && alpha <= 1f, "Alpha component is invalid")

  override def toRGB: RGBColor = {

    def toRGB(r: Float, g: Float, b: Float): RGBColor = {
      RGBColor((r * 255f + 0.5f).toInt, (g * 255f + 0.5f).toInt, (b * 255f + 0.5f).toInt, (alpha * 255f + 0.5f).toInt)
    }

    val h = (hue * 6).toInt
    val f = hue * 6 - h
    val p = value * (1 - saturation)
    val q = value * (1 - f * saturation)
    val t = value * (1 - (1 - f) * saturation)

    h match {
      case 0 => toRGB(value, t, p);
      case 1 => toRGB(q, value, p);
      case 2 => toRGB(p, value, t);
      case 3 => toRGB(p, q, value);
      case 4 => toRGB(t, p, value);
      case 5 => toRGB(value, p, q);
      case _ => throw new RuntimeException(s"Cannot convert from HSV to RGB ($this)")
    }
  }
}

/**
 * The hue component should be between 0.0 and 360.0
 * The saturation component should be between 0.0 and 1.0
 * The lightness component should be between 0.0 and 1.0
 * The alpha component should be between 0.0 and 1.0
 *
 * @param hue
 * @param saturation
 * @param lightness
 */
case class HSLColor(hue: Float, saturation: Float, lightness: Float, alpha: Float) extends Color {
  require(0 <= hue && hue <= 360f, "Hue component is invalid")
  require(0 <= saturation && saturation <= 1f, "Saturation component is invalid")
  require(0 <= lightness && lightness <= 1f, "Lightness component is invalid")
  require(0 <= alpha && alpha <= 1f, "Alpha component is invalid")

  override def toRGB: RGBColor = {
    val h = (hue % 360f) / 360f
    val q = {
      if (lightness < 0.5) lightness * (1 + saturation)
      else (lightness + saturation) - (saturation * lightness)
    }
    val p = 2 * lightness - q
    def hue2rgb(p: Float, q: Float, h: Float): Float = {
      val hprime = {
        if (h < 0) h + 1
        else if (h > 1) h - 1
        else h
      }
      if (hprime * 6 < 1f) p + (q - p) * 6 * hprime
      else if (hprime * 6 < 2f) q
      else if (hprime * 6 < 3f) p + (q - p) * (2f / 3f - hprime) * 6
      else p
    }
    val r = Math.max(0, hue2rgb(p, q, h + (1.0f / 3.0f)))
    val g = Math.max(0, hue2rgb(p, q, h))
    val b = Math.max(0, hue2rgb(p, q, h - (1.0f / 3.0f)))

    RGBColor((r * 255f + 0.5f).toInt, (g * 255f + 0.5f).toInt, (b * 255f + 0.5f).toInt, (alpha * 255f + 0.5f).toInt)
  }
}

File: scrimage-core/src/main/java/com/sksamuel/scrimage/color/RGBColor.java
package com.sksamuel.scrimage.color;

import java.util.Objects;

/**
 * Red/Green/Blue
 * <p>
 * The red, green, blue, and alpha components should be between [0,255].
 */
public class RGBColor implements Color {

   public final int red;
   public final int green;
   public final int blue;
   public final int alpha;

   /**
    * Creates an RGB with default alpha of 255.
    */
   public RGBColor(int red, int green, int blue) {
      this(red, green, blue, 255);
   }

   public RGBColor(int red, int green, int blue, int alpha) {

      assert (0 <= red && red <= 255);
      assert (0 <= green && green <= 255);
      assert (0 <= blue && blue <= 255);
      assert (0 <= alpha && alpha <= 255);

      this.red = red;
      this.green = green;
      this.blue = blue;
      this.alpha = alpha;
   }

   public static RGBColor fromARGBInt(int argb) {
      int alpha = argb >> 24 & 0xFF;
      int red = argb >> 16 & 0xFF;
      int green = argb >> 8 & 0xFF;
      int blue = argb & 0xFF;
      return new RGBColor(red, green, blue, alpha);
   }

   public static RGBColor fromAwt(java.awt.Color color) {
      return new RGBColor(color.getRed(), color.getGreen(), color.getBlue(), color.getAlpha());
   }

   /**
    * Returns as an int the value of this color.
    * The RGB and alpha components are packed as rgba.
    * Use toARGBInt() for clarity.
    */
   @Deprecated
   public int toInt() {
      return toARGBInt();
   }

   public int toARGBInt() {
      return ((alpha & 0xFF) << 24) | ((red & 0xFF) << 16) | ((green & 0xFF) << 8) | ((blue & 0xFF));
   }

   public RGBColor toRGB() {
      return this;
   }

   public RGBColor withRed(int r) {
      return new RGBColor(r, green, blue, alpha);
   }

   public RGBColor withBlue(int b) {
      return new RGBColor(red, green, b, alpha);
   }

   public RGBColor withGreen(int g) {
      return new RGBColor(red, g, blue, alpha);
   }

   public int[] toArray() {
      return new int[]{alpha, red, green, blue};
   }

   /**
    * Returns a conversion of this color into a CYMK color.
    * If this colour is already a CYMK then the same instance will be returned.
    */
   public CMYKColor toCMYK() {
      float max = Math.max(Math.max(red, green), blue) / 255f;
      float k = 1.0f - max;
      if (max > 0.0f) {
         return new CMYKColor(1.0f - (red / 255f) / max, 1.0f - (green / 255f) / max, 1.0f - (blue / 255f) / max, k);
      } else {
         return new CMYKColor(0, 0, 0, k);
      }
   }

   // credit to https://github.com/mjackson/mjijackson.github.com/blob/master/2008/02/rgb-to-hsl-and-rgb-to-hsv-color-model-conversion-algorithms-in-javascript.txt
   @Override
   public HSVColor toHSV() {

      float r = red / 255f;
      float g = green / 255f;
      float b = blue / 255f;

      float max = Math.max(Math.max(r, g), b);
      float min = Math.min(Math.min(r, g), b);

      float d = (max - min);
      float s;
      if (max == 0) {
         s = 0;
      } else {
         s = d / max;
      }

      float h;
      if (max == min) {
         // achromatic
         h = 0f;
      } else {
         float f;
         if (max == r) {
            f = (g - b) / d + (g < b ? 6 : 0);
         } else if (max == g) {
            f = (b - r) / d + 2;
         } else {
            f = (r - g) / d + 4;
         }
         h = f / 6;
      }

      return new HSVColor(h * 360f, s, max, 1f);
   }

   // credit to https://github.com/mjackson/mjijackson.github.com/blob/master/2008/02/rgb-to-hsl-and-rgb-to-hsv-color-model-conversion-algorithms-in-javascript.txt
   public HSLColor toHSL() {

      float r = red / 255f;
      float g = green / 255f;
      float b = blue / 255f;

      float max = Math.max(Math.max(r, g), b);
      float min = Math.min(Math.min(r, g), b);

      float l = (max + min) / 2f;

      float h;
      float s;
      if (max == min) {
         h = 0;
         s = 0;
      } else {
         float d = max - min;

         if (l > 0.5) s = d / (2 - max - min);
         else s = d / (max + min);

         float f;
         if (max == r) {
            f = (g - b) / d + (g < b ? 6 : 0);
         } else if (max == g) {
            f = (b - r) / d + 2;
         } else {
            f = (r - g) / d + 4;
         }
         h = f / 6;
      }

      return new HSLColor(h * 360f, s, l, 1f);
   }

   @Override
   public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      RGBColor rgbColor = (RGBColor) o;
      return red == rgbColor.red &&
         green == rgbColor.green &&
         blue == rgbColor.blue &&
         alpha == rgbColor.alpha;
   }

   @Override
   public int hashCode() {
      return Objects.hash(red, green, blue, alpha);
   }
}
